File: 84.largest-rectangle-in-histogram.cpp
/*
 * @lc app=leetcode id=84 lang=cpp
 *
 * [84] Largest Rectangle in Histogram
 */

// @lc code=start
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        stack<int> st;
        int i = 0;
        int max = 0;
        int n = heights.size();
        while (i <= n) {
            if (st.empty() || (i < n && heights[i] > heights[st.top()])) {
                st.push(i++);
            } else {
                int h = heights[st.top()];
                st.pop();
                int w = st.empty() ? i : (i - 1 - st.top());
                int area = h * w;
                max = area > max ? area : max;
            }
        }

        return max;
    }
};
// @lc code=end

File: 84.largest-rectangle-in-histogram.java
import java.util.Stack;

/*
 * @lc app=leetcode id=84 lang=java
 *
 * [84] Largest Rectangle in Histogram
 */

// @lc code=start
class Solution {
    public int largestRectangleArea(int[] heights) {
        int m = heights.length, index = 0;
        int[] stack = new int[m];
        int max = 0;
        for (int i = 0; i <= m; i++) {
            if (index == 0) {
                stack[index++] = i;
                continue;
            }
            while (index > 0 && (i == m || heights[i] < heights[stack[index - 1]])) {
                int h = heights[stack[--index]];
                int w = index == 0 ? i : i - stack[index - 1] - 1;
                int area = w * h;
                if (area > max) {
                    max = area;
                }
            }
            stack[index++] = i;
        }

        return max;
    }
}
// @lc code=end

